File: turbo.rs
use std::fs::File;
use std::io::{stdout, Write};
use std::os::unix::io::{AsRawFd, FromRawFd};

fn increase_str_num(input: &mut [u8]) -> usize {
    let input_size = input.len();
    for i in (0..input_size).rev() {
        input[i] += 1;
        if input[i] - b'0' < 10 {
            return input_size - i;
        }
        input[i] = b'0';
    }
    panic!("overflowed input");
}

fn main() {
    let stdout = AsRawFd::as_raw_fd(&stdout());
    let mut stdout: File = unsafe {
        FromRawFd::from_raw_fd(stdout)
    };
    const BUFSIZ: usize = 8192;
    let mut buffer = Vec::with_capacity(BUFSIZ);
    let mut num = [b'0';12];
    let mut num_len = 1;
    let prefix = b"Hello, ";
    const LINE_MAX_LEN: usize = 17;
    for _ in 0..100_000_000 {
        buffer.extend_from_slice(prefix);
        buffer.extend_from_slice(&num[num.len() - num_len..]);
        num_len = num_len.max(increase_str_num(&mut num));
        buffer.push(b'\t');
        if buffer.len() + LINE_MAX_LEN > BUFSIZ {
            stdout.write(&buffer).unwrap();
            buffer.clear();
        }
    }
    stdout.write(&buffer).unwrap();
}

File: turbo.c
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
 
#define MAX(a, b) ((a) > (b) ? (a) : (b))
 
int increase_str_num(char *input, int input_size) {
    int i = input_size - 1;
    for(; i >= 0; --i) {
        ++input[i];
        int n = input[i] - '0';
        if (n == 10) input[i] = '0';
        else break;
    }
    return input_size - i;
}
 
int main() {
    char buffer[BUFSIZ];
    char num[12] = "000000000000";
    int num_len = 1;
    const char prefix[] = "Hello, ";
    int buf_index = 0;
    const int line_max_len = 17;
    unsigned int i = 0;
    int n;
 
    int num_len_increase = 1;
    const unsigned int target = 100000000;
 
    while(i < target) {
        while(buf_index < sizeof(buffer) - line_max_len && i < target) {
            memcpy(buffer + buf_index, prefix, sizeof(prefix) - 1);
            buf_index += sizeof(prefix) - 1;
            num_len = MAX(num_len, num_len_increase);
            memcpy(buffer + buf_index, num + sizeof(num) - num_len, num_len);
            buf_index += num_len;
            ++i;
            buffer[buf_index] = '\t';
            ++buf_index;
            num_len_increase = increase_str_num(num, sizeof(num));
        }
        n = write(STDOUT_FILENO, buffer, buf_index);
        if (n == -1) {
            perror("write");
            return -1;
        }
        buf_index = 0;
    }
    return 0;
}
